src/fasta/fasta.cpp

```diff
@@ -36,9 +37,19 @@ ebi::vcf::fasta::FileBasedFasta::FileBasedFasta(const std::string& fasta_path, c
     ebi::util::open_file(fasta_input, fasta_path, std::ifstream::binary);
 
     BOOST_LOG_TRIVIAL(info) << "Reading from input FASTA index file...";
-    std::ifstream fs;
-    ebi::util::open_file(fs, fasta_index_path, std::ifstream::binary);
-    fasta_index = bioio::read_fasta_index(fs);
+    try {
+        std::ifstream fs;
+        ebi::util::open_file(fs, fasta_index_path, std::ifstream::binary);
+        fasta_index = bioio::read_fasta_index(fs);
+    } catch (const std::runtime_error& error) {
+        // Fasta index file does not exist. Create index on-the-fly.
+        BOOST_LOG_TRIVIAL(info) << "Creating index from input FASTA file...";
+        std::stringstream faidx_index;
+        ebi::vcf::faidx::index_fasta(fasta_input, faidx_index);
+        fasta_input.clear();
+        fasta_input.seekg(0, std::ios::beg);
```

Comment:
Given that these 2 actions need to be happen for the index to be read properly after creation, maybe it would make sense to reset the stream at the end of the `index_fasta` method? That way it will always be created and leave everything ready.

Reply:
In `index_fasta`, the reference is `std::ostream` type so you can only set the pointer for `put` position through `seekp` call. That would have no effect on the `get` position pointer if these two pointers are different as in most (if not all) implementations.

However I am told that whether to use two separate pointers for `get` and `put` is implementation dependent. In the case in which the same pointer is used for `get` and `put` these two lines would be needed here. In most cases in which two separate pointers are used, these two lines are not necessary, but I would like to leave them here to make the intention clear.